pkg/agent/globalstate.go

```diff
@@ -172,9 +146,167 @@ func (s *agentState) handleVMEventAdded(
+		klog.Errorf("TriggerRestartIfNecessary called with zero'd Time for pod %v", podName)
+		// Continue on, but with the time overridden, so we guarantee our minimum wait.
+		endTime = time.Now()
+	}
+
+	// keep this for later.
+	exitKind := pod.status.endState.ExitKind
+
+	switch exitKind {
+	case podStatusExitCanceled:
+		return // successful exit, no need to restart.
+	case podStatusExitPanicked, podStatusExitErrored:
+		// Should restart; continue.
+	default:
+		klog.Errorf("TriggerRestartIfNecessary called with unexpected ExitKind %q for pod %v", podName)
+		return
+	}
+
+	// Begin steps (2) and (3) -- wait, then restart.
+	var waitDuration time.Duration
+	totalRuntime := endTime.Sub(pod.status.startTime)
+
+	// If the runner was running for a while, restart immediately.
+	//
+	// NOTE: this will have incorrect behavior when the system clock is behaving weirdly, but that's
+	// mostly ok. It's ok to e.g. restart an extra time at the switchover to daylight saving time.
+	if totalRuntime > time.Second*time.Duration(RunnerRestartMaxWaitSeconds) {
+		waitDuration = 0
+	} else /* Otherwise, randomly pick within RunnerRestartMinWait..RunnerRestartMaxWait */ {
+		r := util.NewTimeRange(time.Second, RunnerRestartMinWaitSeconds, RunnerRestartMaxWaitSeconds)
+		waitDuration = r.Random()
+		klog.Info("Waiting for %s before restarting Runner %v", waitDuration, podName)
+	}
+
+	// Run the waiting (if necessary) and restarting in another goroutine, so we're not blocking the
+	// caller of this function.
+	go func() {
+		if waitDuration != 0 {
+			select {
+			case <-time.After(waitDuration):
+			case <-runnerCtx.Done():
+				klog.Infof("Canceling restart of Runner %v after %s: %s", podName, time.Since(endTime), runnerCtx.Err())
+				return
+			}
+		}
+
+		s.lock.Lock()
+		defer s.lock.Unlock()
+
+		pod, ok := s.pods[podName]
+		if !ok {
+			klog.Warningf("Canceling restart of Runner %v after %s: no longer present in pod map", podName, time.Since(endTime))
+			return
+		}
+
+		pod.status.mu.Lock()
+		defer pod.status.mu.Unlock()
+
+		// Runner was already restarted
+		if pod.status.endState == nil {
+			addedInfo := "this generally shouldn't happen, but could if there's a new pod with the same name"
+			klog.Warningf(
+				"Canceling restart of Runner %v after %s: Runner was already restarted (%s)",
+				podName, time.Since(endTime), addedInfo,
+			)
+		}
+
+		klog.Infof("Restarting %s Runner %v after %s, was running for %s", exitKind, podName, time.Since(endTime), totalRuntime)
+		//                     ^^
+		// note: exitKind is one of "panicked" or "errored" - e.g. "Restarting panicked Runner ..."
+
+		runner := s.newRunner(pod.status, pod.status.vmInfo, podName, podIP)
+
+		txVMUpdate, rxVMUpdate := util.NewCondChannelPair()
+		// note: pod is *podState, so we don't need to re-assign to the map.
+		pod.vmInfoUpdated = txVMUpdate
+		pod.runner = runner
+
+		pod.status.previousEndStates = append(pod.status.previousEndStates, *pod.status.endState)
```

Comment:
is there a cap on how long we're willing to let this slice grow?

Reply:
Not currently, no. It should only increase every 5s, so IMO that's probably fine.

(if we're getting OOM-killed because things are restarting too much, then we probably have bigger issues)

Reply:
I mean, if runners are going to crash (and all of them might at the same time if there's a bug? possibly?) and that happens a lot... hitting memory limits seems quite possible, but I suppose there's not much to do other than panic, and the oom will do the thing.

Because I don't have a great mental model of this: do these states end up in the "dump state endpoint", so if this got big would that endpoint die?

Reply:
> suppose there's not much to do other than panic, and the oom will do the thing

:+1:

> do these states end up in the "dump state endpoint", so if this got big would that endpoint die?

Realistically I think this would have to get _really_ big for that to happen — the only processing we do on these is to copy it into a new slice, then convert to JSON and respond.

We have metrics for runner panics (and #274 adds for fatal errors), so we should be getting a consistent "SYSTEM IS ON FIRE" signal long before that happens.